p2p/src/sync/peer_v2/transaction_manager.rs

```diff
@@ -150,6 +164,8 @@ where
                 _ = maintenance_interval.tick() => {}
             }
 
+            self.announce_transactions_if_needed()?;
```

Comment:
Why not use proper futures with await to do the delays, instead of this ad-hoc way of calculating times?

Reply:
I'm not sure what you mean exactly by proper solution. There is an event loop per peer that we utilise to handle events and execute code.

Reply:
Maybe I'm missing something, but let's see. Please discuss this with Mykhailo and Lukas and see whether they know how to exactly do this since they dealt with async more than I did.

Basically your event loop suspends at `select!`, and pops a future that's ready out of the different arms. What you're doing is ignoring that, and only after an event happens, manually check whether "it's time to send a transaction".

I believe there's a better way.

First, you have an async tokio sleep function that returns a future that can be an arm there. That's the first component in this puzzle. So, assuming we have only one transaction that you want to send out, all you have to do is await on the sleep future in that `select!` for that transaction.

But now we need to figure this out for multiple transactions. You need some kind of async iterator that iterates over elements of a container that contains all the transactions to be sent out, where every time a future is ready, it gets popped out of the container. Worst case scenario, you can create your own type and implement Future for that. However, I'm sure there's a way that wouldn't require you to do that manually since this is a typical use case. There's a `join!` macro for futures, but you want the opposite. You want something that takes a set of futures, and you should be able to pop elements out of it when they're ready. Not sure what that is. I'm sure we used that somewhere in the code, perhaps in the peers event loop to receive messages.

Now that would be a proper use of an async event loop. Hopefully I'm not missing something.

Reply:
> But now we need to figure this out for multiple transactions. You need some kind of async iterator that iterates over elements of a container that contains all the transactions to be sent out, where every time a future is ready, it gets popped out of the container. Worst case scenario, you can create your own type and implement Future for that. However, I'm sure there's a way that wouldn't require you to do that manually since this is a typical use case. There's a `join!` macro for futures, but you want the opposite. You want something that takes a set of futures, and you should be able to pop elements out of it when they're ready. Not sure what that is. I'm sure we used that somewhere in the code, perhaps in the peers event loop to receive messages.

The way you can go about this is to have a data structure collecting delayed set of actions. It could be something like a priority queue that returns the elements in the order of the delay. Also introduce a member function (let's call it `due`) that just sleeps for the amount of time needed for the first element to become due, or suspends (`future::pending`) if there's nothing. No need for a mechanism to wait on multiple futures since by keeping the elements in the priority queue, you can easily access the element that is supposed to be processed first and wait for that one. It could even be generic, parametrized over data it contains (transactions in this case), and chuck it in `utils`, I suspect it's useful for other things too.

Reply:
Thanks for the details; that's very close to how I pictured it. I'm just hoping someone already created this because, in my head, it seems like a common thing to need. We should look online and try to find it.

Reply:
looks like `FuturesUnordered` is the standard tool for that

Reply:
Sounds about right. I do remember we used this already elsewhere in the code. Cheers!

Reply:
Yea `FuturesUnordered` can work but may be too generic. It will poll all the futures in the set. However in this case you know the order in which the futures will resolve so it's sufficient to pool only the most immediate one. With `FuturesUnordered` the cost model (and by implication the security model with regards to DoS attacks) will be a bit more complicated (maybe not too bad since all transactions still have to pass the min fee check). Seems to me turning `O(n)` polls into `O(1)` is worth it since the additional complexity is not massive compared to the off the shelf solution. Remember the polls will be performed on every iteration of the event loop.

Reply:
Two things:

1. This might be possible if it's possible to poll the remaining time from the future, which is not really that simple unless we want to recreate the adhoc problem. I might be wrong there though.
2. I'm OK with FuturesUnordered now, at least as a first go. We can create an issue and fix it later to make it idea. Spending days to create the ideal solution isn't really a priority now. I don't see how any of this can lead to DoS attacks, as long as the size of that container is limited.

Reply:
> 1. This might be possible if it's possible to poll the remaining time from the future, which is not really that simple unless we want to recreate the adhoc problem. I might be wrong there though.

It's not necessary to query the remaining time, just create the future as needed. Rough sketch:
```rust
loop {
    select! {
        // Other event loop stuff HERE
        _ = sleep(pending_txs.time_left_till_first_one()) => process_tx(pending_txs.pop()),
    }
}
```

Of course it should be properly encapsulated. Possibly the future for the first item deadline may be cached.

Reply:
take a look guys